### PR TITLE
feat: client fingerprint usecase

### PR DESCRIPTION
**Pull request by MohamadJaara**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

we need to show client Finger print

### Solutions

ClientFingerprintUseCase that does the following 
1. get client fingerprint from its session
2. if success returns.
3. if SESSION_NOT_FOUND error, then establish session and try again.
4. if any other error return.

In order to achieve step 3 i have to do some refactoring and move the session establish logic form the sessionEstablisher to the prekeyRepository so we can share the same logic in multiple places

### Dependencies (Optional)

_If there are some other pull requests related to this one (e.g. new releases of frameworks), specify them here._

Needs releases with:

- [ ] GitHub link to other pull request

### Testing

Unit test added

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: cryptography/src/androidMain/kotlin/com/wire/kalium/cryptography/ProteusClientCryptoBoxImpl.kt [hunk 1]
```diff
@@ -54,6 +54,10 @@ class ProteusClientCryptoBoxImpl constructor(
         return false
     }
 
+    override suspend fun remoteFingerPrint(sessionId: CryptoSessionId): ByteArray = withContext(defaultContext) {
+        wrapException { box.getSession(sessionId.value).remoteFingerprint }
+    }
```
Comment (vitorhugods) [inline cryptography/src/androidMain/kotlin/com/wire/kalium/cryptography/ProteusClientCryptoBoxImpl.kt:59]: Cover with tests?

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/data/prekey/PreKeyRepositoryTest.kt [hunk 2]
```diff
@@ -111,6 +128,128 @@ class PreKeyRepositoryTest {
             .wasInvoked(exactly = once)
     }
 
+    @Test
+    fun givenFetchingPreKeysWithNullClients_whenPreparingSessions_thenTryToInvalidateINvalidSessions() = runTest {
```
Comment (vitorhugods) [inline logic/src/commonTest/kotlin/com/wire/kalium/logic/data/prekey/PreKeyRepositoryTest.kt:132]:
```suggestion
    fun givenFetchingPreKeysWithNullClients_whenPreparingSessions_thenTryToInvalidateInvalidSessions() = runTest {
```

File: logic/src/commonTest/kotlin/com/wire/kalium/logic/data/prekey/PreKeyRepositoryTest.kt [hunk 3]
```diff
@@ -111,6 +128,128 @@ class PreKeyRepositoryTest {
...
+        val preKey = PreKeyDTO(42, "encodedData")
+        val prekeyCrypto = PreKeyCrypto(preKey.id, preKey.key)
+
+        val userPreKeysResult = NetworkResponse.Success(
+            mapOf(
+                TEST_USER_ID_1.domain to mapOf(
+                    TEST_USER_ID_1.value to mapOf(
+                        TEST_CLIENT_ID_1.value to preKey,
+                        "invalidClient" to null,
+
```
Comment (vitorhugods) [inline logic/src/commonTest/kotlin/com/wire/kalium/logic/data/prekey/PreKeyRepositoryTest.kt:142]:
```suggestion
```

File: cryptography/src/jvmMain/kotlin/com/wire/kalium/cryptography/ProteusClientCryptoBoxImpl.kt [hunk 4]
```diff
@@ -84,6 +84,10 @@ class ProteusClientCryptoBoxImpl constructor(
         return wrapException { box.localFingerprint }
     }
 
+    override suspend fun remoteFingerPrint(sessionId: CryptoSessionId): ByteArray {
+        TODO("Not yet implemented")
```
Comment (vitorhugods) [inline cryptography/src/jvmMain/kotlin/com/wire/kalium/cryptography/ProteusClientCryptoBoxImpl.kt:88]: We can implement this in https://github.com/wireapp/cryptobox4j.

But at the same time... We're dropping Cryptobox for CoreCrypto anyway, so I guess it's fine to leave it like this.

**Review by vitorhugods — APPROVED** [review]
Nicely done.

Would be great to cover the `remoteFingerPrint` function with tests.

Other than that, just a couple of nitpicks.

File: cryptography/src/androidMain/kotlin/com/wire/kalium/cryptography/ProteusClientCryptoBoxImpl.kt (hunk 1 above)
Comment (MohamadJaara) [reply cryptography/src/androidMain/kotlin/com/wire/kalium/cryptography/ProteusClientCryptoBoxImpl.kt:59]: done

**Pull request merged**